### ISSUE TITLE
Sync not deleting local files

### ISSUE THREAD

**Issue by chasebolt** [devspace-cloud/devspace#903]
**What happened?**  
I removed the composer package `phpunit` from the `composer.json`.

**What did you expect to happen instead?**  
I expected the local synced copy of `/vendor` to no longer contain `/vendor/phpunit` after redeploying. `composer install` is part of the `Dockerfile`. The file does not exist in the container (as expected), but it remains in my local environment.

**How can we reproduce the bug?** (as minimally and precisely as possible)  
```yaml
sync:
  - downloadOnInitialSync: true
    uploadExcludePaths:
      - vendor
```

**Local Environment:**  
- Operating System: linux
- Deployment method: helm

**Kubernetes Cluster:**  
- Cloud Provider: minikube
- Kubernetes Version: 1.17.2

/kind bug

**Comment by FabianKramm** [comment]
@chasebolt thanks for opening this issue! Does it work when you change the `uploadExcludePaths` to `excludePaths`?

**Comment by chasebolt** [comment]
i do not have a vendor folder locally so it will delete the remote vendor folder if i remove uploadExcludePaths.

**Comment by chasebolt** [comment]
the vendor folder is created by composer install during the docker build. my users don't have composer installed on their workstation, but they do need this vendor folder synced locally for their IDE.

ideally it would be nice if we could have a flag for sync (if this isn't implemented in a way already) to say which side (local/remote) is preferred. i would flag the remote vendor directory is preferred which then would sync locally and also keep it as a bidirectional sync. uploadExcludePaths breaks bidirectional, which is a workaround for my case, but not ideal.

**Comment by chasebolt** [comment]
ive also tried it with just 

```yaml
sync:
  - downloadOnInitialSync: true
```

and it does not remove local files that do not exist on the remote. it will end up syncing the local files up to the remote.

**Comment by FabianKramm** [comment]
@chasebolt thanks for the information! The option `downloadOnInitialSync` means that initially files are downloaded. So I tried this with the quickstart example and it works for me as expected with the following configuration in local docker kubernetes / gke:

```
version: v1beta5
images:
  default:
    image: dscr.io/${DEVSPACE_USERNAME}/devspace
deployments:
- name: quickstart
  helm:
    componentChart: true
    values:
      containers:
      - image: dscr.io/${DEVSPACE_USERNAME}/devspace
      service:
        ports:
        - port: 3000
dev:
  ports:
  - imageName: default
    forward:
    - port: 3000
  autoReload:
    paths:
    - package.json
    - Dockerfile
  sync:
  - imageName: default
    downloadOnInitialSync: true
    uploadExcludePaths:
    - node_modules/
```

Then I run `devspace dev` and after everything is synced I delete a folder in the `node_modules` and the folder is also deleted locally. However, if content is newer locally then on the remote side, files will not be deleted. Could you send the contents of the .devspace/logs/sync.log file here, you should see notifications if files were skipped during deletion?

Could you also try with this config & the [quickstart example](https://github.com/devspace-cloud/devspace/tree/master/examples/quickstart)?

**Comment by chasebolt** [comment]
the issue comes up when you stop `devspace dev`, remove a package from composer.json, and then run `devspace dev` again. the package that was removed wont exist on the remote, wont be deleted locally, and then the sync will sync the local package to the remote.

**Comment by FabianKramm** [comment]
@chasebolt yes you are correct, the documentation states that this is possible, however currently it isn't. The only workaround I currently see is deleting the folder on every `devspace dev` via a hook. However, we will definitively fix this in the next version.

**Comment by chasebolt** [comment]
Thanks for the suggested workaround!

**Comment by FabianKramm** [comment]
@chasebolt After some discussion, we came up with the following format: we will introduce a new sync property called `initialSync` that specifies the initial sync strategy devspace will do. There are 4 different options:
- mirrorLocal: basically mirrors the local state onto the remote state (overrides outdated files and deletes files remotely that are not there locally, the current default) 
- mirrorRemote: the same as mirrorLocal, but just the other way around
- preferLocal: does not delete any files, but overrides outdated remote files
- preferRemote: same as preferLocal, but prefers remote files over local files

In your case you would just create 2 sync configs, one for the source folder with the default strategy and one for the vendor folder with the mirrorRemote strategy, which should solve your problem. I'll add this to my todo list and it should be done for the next release.

**Comment by chasebolt** [comment]
Thank you, that sounds like an elegant solution to this issue. 👍

**Comment by FabianKramm** [comment]
Implemented with #983
